sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala

```diff
@@ -1312,6 +1314,109 @@ class SumBinaryFixer(toType: DataType, isAnsi: Boolean)
   }
 }
 
+/**
+ * Fixes up a sum operation for unbounded preceding to unbounded following
+ * @param errorOnOverflow if we need to throw an exception when an overflow happens or not.
+ */
+class SumUnboundedToUnboundedFixer(resultType: DataType, failOnError: Boolean)
+    extends BatchedUnboundedToUnboundedWindowFixer {
+
+  private var previousValue: Option[Scalar] = None
+
+  private var hasOverflowed = false
+
+  private val numeric = TypeUtils.getNumeric(resultType, failOnError)
+
+  override def updateState(scalar: Scalar): Unit = {
+    if (hasOverflowed) {
+      return
+    }
+    previousValue match {
+      case None =>
+        previousValue = Some(scalar.incRefCount())
+      case Some(prev) if scalar.isValid =>
+        withResource(prev) { _ =>
+          previousValue = None
+          prev.getType.getTypeId match {
+            case DType.DTypeEnum.INT64 =>
+              if (failOnError) {
+                previousValue = Some(Scalar.fromLong(Math.addExact(
+                  scalar.getLong, prev.getLong)))
+              } else {
+                previousValue = Some(Scalar.fromLong(scalar.getLong + prev.getLong))
+              }
+            case DType.DTypeEnum.FLOAT32 =>
+              previousValue = Some(Scalar.fromFloat(scalar.getFloat + prev.getFloat))
+            case DType.DTypeEnum.FLOAT64 =>
+              previousValue = Some(Scalar.fromDouble(scalar.getDouble + prev.getDouble))
+            case DType.DTypeEnum.DECIMAL32 | DType.DTypeEnum.DECIMAL64 |
+                 DType.DTypeEnum.DECIMAL128 =>
+              val decimal = numeric.plus(Decimal(prev.getBigDecimal),
+                Decimal(scalar.getBigDecimal)).asInstanceOf[Decimal]
+              val dt = resultType.asInstanceOf[DecimalType]
+              previousValue = Option(TrampolineUtil.checkDecimalOverflow(
```

Comment:
I did some more testing and unfortunately the overflow is not sticky for decimal values. I think we are going to need to either fallback to the old way for Decimals in non-ANSI mode or find a way to know that an overflow happened both within a batch and make the overflow between batches sticky.

Say I have two batches of input. The first batch overflows, and the result is a null, but the second batch does not. Now we try to add null to something that is not null. And we get the not-null value back. But in reality the overflow should be sticky. Once we overflow we can never go back.

Reply:
I had run into this issue during testing and added the `hasOverflowed` to handle that. Is the issue around how this is distributed? Are there multiple instances of `SumUnboundedToUnboundedFixer` involved and that is why we can't catch all cases?

Reply:
On Spark 3.3.0
```
spark.conf.set("spark.rapids.sql.batchSizeBytes", "100")
spark.time(spark.range(20).repartition(1).selectExpr("id as a", "if(id = 0, 99999999999999999999999999999999999999, 1) as b").selectExpr("SUM(b) OVER (ORDER BY a ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) as all_sum_b","*").show(false))
```

But out of the box this is exposing some other bugs. To fix it I had to apply the following patch.

```diff
diff --git a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala
index c96f41400..d13777e4e 100644
--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExpression.scala
@@ -1356,7 +1356,7 @@ class SumUnboundedToUnboundedFixer(resultType: DataType, failOnError: Boolean)
               val dt = resultType.asInstanceOf[DecimalType]
               previousValue = Option(TrampolineUtil.checkDecimalOverflow(
                   decimal, dt.precision, dt.scale, failOnError))
-                .map(n => Scalar.fromDecimal(n.toJavaBigDecimal))
+                .map(n => GpuScalar.from(n, dt))
               if (previousValue.isEmpty) {
                 hasOverflowed = true
               }
```

The problem is that the overflow code only detects an overflow if it happened in between batches. But if the overflow happens within a batch it has no knowledge of that and assumes the null returned is the same as if all the data in that batch were nulls.

Reply:
I applied this suggested change but was not able to reproduce the issue locally.

Reply:
```diff
diff --git a/integration_tests/src/main/python/window_function_test.py b/integration_tests/src/main/python/window_function_test.py
index 339954177..39d965419 100644
--- a/integration_tests/src/main/python/window_function_test.py
+++ b/integration_tests/src/main/python/window_function_test.py
@@ -292,6 +292,35 @@ def test_numeric_sum_window_unbounded(data_gen, partition_by):
         conf = {'spark.rapids.sql.variableFloatAgg.enabled': 'true',
                 'spark.rapids.sql.batchSizeBytes': '100'})
 
+@ignore_order
+@pytest.mark.parametrize('batch_size', ['100', '1g'], ids=idfn)
+def test_numeric_sum_window_unbounded_decimal_overflow(batch_size):
+    assert_gpu_and_cpu_are_equal_sql(
+        # the 38 9s are the maximum value that a Decimal(38,0) can hold and Spark will infer that type automaticllay
+        # so the first batch will overflow, if it has at least two rows in it. This verifies that subsiquent batches
+        # can detect the overflow in the first batch and also include that in later batches.
+        lambda spark: spark.range(1024).selectExpr("id as a", "if (id = 0, 99999999999999999999999999999999999999, 1) as b"),
+        'window_agg_table',
+        'select '
+        ' sum(b) over '
+        '   (order by a asc '
+        '      rows between UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) as sum_b_asc '
+        'from window_agg_table',
+        conf = {'spark.rapids.sql.batchSizeBytes': batch_size})
+
+@ignore_order
+@pytest.mark.parametrize('batch_size', ['100', '1g'], ids=idfn)
+def test_numeric_sum_window_unbounded_long_overflow(batch_size):
+    assert_gpu_and_cpu_are_equal_sql(
+        lambda spark: spark.range(1024).selectExpr("id as a", "if (id = 0, 9223372036854775807, 1) as b"),
+        'window_agg_table',
+        'select '
+        ' sum(b) over '
+        '   (order by a asc '
+        '      rows between UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) as sum_b_asc '
+        'from window_agg_table',
+        conf = {'spark.rapids.sql.batchSizeBytes': batch_size})
+
 @pytest.mark.xfail(reason="[UNSUPPORTED] Ranges over order by byte column overflow "
                           "(https://github.com/NVIDIA/spark-rapids/pull/2020#issuecomment-838127070)")
 @ignore_order
```

Here is a new patch that adds a test that fails.

Reply:
Thanks. I confirmed that `test_numeric_sum_window_unbounded_decimal_overflow` fails for me with the fix reverted. However, it also fails when the fix is reinstated due to different output between CPU and GPU where the CPU is producing None and the GPU is producing a value, so there is still an issue, but at least I have a failing test for it now.